File: irredloop.py
import random

EXIT, LOOP, FORWARD = 0, 1, 2
def test_irreducible():
    ilt = (irreducible_loops_no_goto,
        irreducible_loops)
    def doinc(z): z[0] += 1
    for i in range(2):
        u, v, w, x = [0], [0], [0], [0]
        random.seed(5)
        ilt[i](random.randint(0, 1) == 1, lambda: random.randint(0, 2),
            lambda: random.randint(0, 2), lambda: doinc(u),
            lambda: doinc(v), lambda: doinc(w), lambda: doinc(x))
        print(u, v, w, x)

def irreducible_loops_no_goto(u_start,
    w_cond, x_cond, u, v, w, x):
    is_end, v_first = [False], not u_start
    def doExit(): is_end[0], loop[0] = True, True
    def doLoop(): loop[0] = True
    def doForward(): pass
    while True:
        if not v_first: u()
        while True:
            if not v_first:
                w()
                loop = [False]
                [doExit, doLoop, doForward][w_cond()]()
                if loop[0]: break
            while True:
                if not v_first:
                    x()
                    loop = [False]
                    [doExit, doLoop, doForward][x_cond()]()
                    if loop[0]: break
                v()
                v_first = False
            if is_end[0]: break
        if is_end[0]: break

JUMP, CONDJUMP, JUMPTABLE, CALLLAMBDA = 0, 1, 2, 3 #opcodes
def irreducible_loops(u_start,
    w_cond, x_cond, u, v, w, x):
    do_u, do_w, do_x, do_v, end = 1, 2, 4, 6, 8
    code = [(CONDJUMP, not u_start, do_v), #0
            (CALLLAMBDA, u), #1=do_u
            (CALLLAMBDA, w), #2=do_w
            (JUMPTABLE, w_cond, [end, do_u, do_x]), #3
            (CALLLAMBDA, x), #4=do_x
            (JUMPTABLE, x_cond, [end, do_w, do_v]), #5
            (CALLLAMBDA, v), #6 = do_v
            (JUMP, do_x)] #7, end is 8
    instptr = 0
    while True:
      if code[instptr][0] == JUMP:
        instptr = code[instptr][1]
      elif code[instptr][0] == CONDJUMP:
        if code[instptr][1]:
          instptr = code[instptr][2]
        else: instptr += 1
      elif code[instptr][0] == JUMPTABLE:
        instptr = code[instptr][2][code[instptr][1]()]
      elif code[instptr][0] == CALLLAMBDA:
        code[instptr][1]()
        instptr += 1
      else: assert False
      if instptr >= len(code): break

test_irreducible()

File: cpp/selfmodgraph/IrredLoop.cpp
#include <functional>
#include <iostream>

//https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.35.8991&rep=rep1&type=pdf
#define EXIT 0
#define LOOP 1
#define FORWARD 2

typedef void(*irred_loop_type)(bool u_start,
    std::function<int()> w_cond, std::function<int()> x_cond,
    std::function<void()> u, std::function<void()> v,
    std::function<void()> w, std::function<void()> x);

void irreducible_loops_no_goto(bool u_start,
    std::function<int()> w_cond, std::function<int()> x_cond,
    std::function<void()> u, std::function<void()> v,
    std::function<void()> w, std::function<void()> x)
{
    bool is_end = false, v_first = !u_start;
    do {
        if (!v_first) u();
        do {
            if (!v_first) {
                w();
                bool loop = false;
                switch (w_cond()) {
                case EXIT: is_end = true;
                case LOOP: loop = true;
                }
                if (loop) break;
            }
            do {
                if (!v_first) {
                    x();
                    bool loop = false;
                    switch (x_cond()) {
                    case EXIT: is_end = true;
                    case LOOP: loop = true;
                    }
                    if (loop) break;
                }
                v();
                v_first = false;
            } while (true);
        } while (!is_end);
    } while (!is_end);
}

void irreducible_loops(bool u_start,
       std::function<int()> w_cond, std::function<int()> x_cond,
       std::function<void()> u, std::function<void()> v,
       std::function<void()> w, std::function<void()> x)
{
    bool loop;
    if (!u_start) goto do_v;
    do {
        u();
        do {
            w();
            loop = false;
            switch (w_cond()) {
            case EXIT: goto end;
            case LOOP: loop = true;
            }
            if (loop) break;
            do {
                x();
                loop = false;
                switch (x_cond()) {
                case EXIT: goto end;
                case LOOP: loop = true;
                }
                if (loop) break;
do_v:
                v();
            } while (true);
        } while (true);
    } while (true);
end:
    return;
}

void irreducible_loops_only_goto(bool u_start,
    std::function<int()> w_cond, std::function<int()> x_cond,
    std::function<void()> u, std::function<void()> v,
    std::function<void()> w, std::function<void()> x)
{
    if (!u_start) goto do_v;
do_u:
    u();
do_w:
    w();
    switch (w_cond()) {
    case EXIT: goto end;
    case LOOP: goto do_u;
    }
do_x:
    x();
    switch (x_cond()) {
    case EXIT: goto end;
    case LOOP: goto do_w;
    }
do_v:
    v();
    goto do_x;
end:
    return;
}

void test_irreducible()
{
    irred_loop_type ilt[] = { irreducible_loops_no_goto,
        irreducible_loops, irreducible_loops_only_goto };
    for (int i = 0; i < 3; i++) {
        int u = 0, v = 0, w = 0, x = 0;
        srand(1);
        ilt[i](rand() & 1, []() { return rand() % 3; },
            []() { return rand() % 3; }, [&u]() { u++; },
            [&v]() { v++; }, [&w]() { w++; }, [&x]() { x++; });
        std::cout << u << " " << v << " " << w << " " << x << std::endl;
    }
}
